README.md
## LAKGVES
[![Build Status](https://travis-ci.org/timgabets/lakgves.svg?branch=master)](https://travis-ci.org/timgabets/lakgves)
[![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)

lak·ves
1. dextral muscles around the eyes, in particular the temporal lobes of the chin and lower lip. (*"the lakgves of other vertebrates"*)
2. [a word that does not exist](https://l.thisworddoesnotexist.com/Wvn4); it was invented, defined and used by a machine learning algorithm.

### Concept

HTTP to [ISO8583](https://en.wikipedia.org/wiki/ISO_8583) message converter, written from scratch in Rust and heavily inspired by [JUKS' socket queue](https://github.com/juks/iso-8583-socket-queue). The main purpose of the system is testing Issuer Bank systems.

                               +-----------------+                              +--------------------+
                               |                 | <------ XML ISO 8583 ------> |    Issuer Bank     |
    HTTP client <--- JSON ---> |     Lakgves     | <--- TODO (VISA SMS/DMS) --> |   Authorization    |
                               |                 | <---  TODO (Mastercard) ---> |        Host        |
                               +--------+--------+                              +--------------------+
                                        |              +--------------------+
                                        |              |    Issuer Bank     |
                                        +-- SP XML --> |     Anti-Fraud     |
                                                       |       System       |
                                                       +--------------------+

The application converts JSON payload to ISO8583 message in a proper format, sends it to the Bank host, receives the response and sends it back to the HTTP client.
In other words, the input JSON like this (note that the card number in i002 and expiration date in i014 are masked here for security reasons. The real request should contain the full unmacked data):
```json
{
  "i000":"0100",
  "i002":"443322******0961",
  "i003":"300000",
  "i004":"000000000000",
  "i014":"****",
  "i018":"6011",
  "i022":"0100",
  "i023":"000",
  "i025":"02",
  "i026":"04",
  "i032":"999999",
  "i041":"TERMID01",
  "i042":"IDDQD MERCH ID",
  "i043":"IDDQD AM. 341215574     341215574 MSKRU",
  "i049":"643",
  "i053":"9801100001000000"
}
```
will be converted into one of the [ISO8583](https://en.wikipedia.org/wiki/ISO_8583) flavours (e.g. Debit Host Interface XML in the example below):
```xml
<RequestInput>
  <ISO8583-87>
    <i000>0100</i000>
    <i002>443322******0961</i002>
    <i003>300000</i003>
    <i004>000000000000</i004>
    <i007>0505151908</i007> <!-- Transmission date & time - generated by the application if not provided -->
    <i011>367298</i011> <!-- STAN - generated by the application if not provided -->
    <i012>151908</i012> <!-- hhmmss - generated by the application if not provided -->
    <i013>0505</i013> <!-- MMDD - generated by the application if not provided -->
    <i014>****</i014>
    <i018>6011</i018>
    <i022>0100</i022>
    <i023>000</i023>
    <i025>02</i025>
    <i026>04</i026>
    <i032>999999</i032>
    <i037>964800787298</i037> <!-- RRN (Retrieval Reference Number) - generated by the application if not provided -->
    <i041>TERMID01</i041>
    <i042>IDDQD MERCH ID</i042>
    <i043>IDDQD AM. 341215574     341215574 MSKRU</i043>
    <i049>643</i049>
    <i053>9801100001000000</i053>
  </ISO8583-87>
</RequestInput>
```


Please refer to your [ISO8583](https://en.wikipedia.org/wiki/ISO_8583) specification for the fields meaning and possible values.


### Supported formats
The list of supported formats:
* [TSYS DHI (Device Host Interface) XML](https://github.com/timgabets/dhi-xml)
* [IBM Safer Payments XML](https://github.com/timgabets/sp-xml)

### Installation
#### Red Hat/Fedora:
Quick install:
```bash
curl -s https://packagecloud.io/install/repositories/timgabets/lakgves/script.rpm.sh | sudo bash
sudo yum install lakgves
```

### Running
Start app as a service:
```bash
sudo systemctl start lakgves
sudo systemctl status lakgves
```

The application logs may be retrived by journalctl:
```
journalctl -u lakgves
```

Alternatively, execute application manually:
```bash
# Edit configuration file /etc/lakgves/config.toml and then run
/usr/bin/lakgves /etc/lakgves/config.toml
```

### Usage
Any HTTP tool (e.g. [curl](https://curl.haxx.se/) or [SoapUI](https://www.soapui.org/)) may be used to interact with the Lakgves app:
```
curl -v --header "Content-Type: application/json" --request POST --data '{"i000":"0100", "i002":"443322******0961", "i003":"300000", "i004":"000000000000", "i014":"****", "i018":"6011", "i022":"0100", "i023":"000", "i025":"02", "i026":"04", "i032":"999999", "i041":"TERMID01", "i042":"IDDQD MERCH ID", "i043":"IDDQD AM. 341215574     341215574 MSKRU", "i049":"643", "i053":"9801100001000000"}' localhost:8080
```
The received response may look like this:
```
< HTTP/1.1 200 OK
< content-length: 528
< content-type: application/json
< x-hdr: sample
< date: Tue, 05 May 2020 15:01:52 GMT
<
* Connection #0 to host localhost left intact
{"i000":"0110","i002":"443322******0961","i003":"300000","i004":"000000000000","i007":"0505180152","i011":"233005","i012":"180152","i013":"0505","i014":"****","i018":"6011","i022":"0100","i023":"000","i025":"02","i026":"4","i032":"999999","i037":"634538143910","i038":"082673","i039":"00","i041":"TERMID01","i042":"IDDQD MERCH ID","i043":"IDDQD AM. 341215574     341215574 MSKRU","i049":"643","i053":"9801100001000000","i054":"0001643D000001431065","i096":"0000634538143910","i120":"UD038IR0044444CR009ES0048100IA0103510301943"}* Closing connection 0
curl -v --header "Content-Type: application/json" --request POST --data
```

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.18] - 2020-07-08
### Added
- Using extfg-sigma version with fixed auth serno overflowing bug.

## [0.1.17] - 2020-07-07
### Added
- extfg-sigma message protocol support.

## [0.1.16] - 2020-06-15
### Added
- Managing application with systemd's ``systemctl``.

## [0.1.15] - 2020-06-10
### Added
- Building RPM package with ``cargo rpm build``.

## [0.1.14] - 2020-06-04
### Changed
- Using AtomicUsize instead of Mutex to switch between multiple connections.

## [0.1.13] - 2020-06-04
### Added
- External library for TSYS DHI XML message protocol serialization/deserialization.

## [0.1.12] - 2020-06-02
### Added
- External library for IBM Safer Payments® XML message protocol serialization/deserialization.

## [0.1.11] - 2020-06-01
### Added
- IBM Safer Payments® XML dummy message protocol support.

## [0.1.10] - 2020-05-11
### Added
- Returning 504 Gateway Timeout if there is 5 second timeout occurred while communicating with DHI host.

## [0.1.9] - 2020-05-05
### Added
- Parsing .toml configuration file.

## [0.1.8] - 2020-05-05
### Added
- Parsing command-line options (configuration file, help, version).

## [0.1.7] - 2020-05-04
### Changed
- Using actix-web framework.

## [0.1.6] - 2020-04-28
### Added
- DHI response deserialization.
- Converting DHI response to JSON and sending it back as HTTP payload.

## [0.1.5] - 2020-04-27
### Changed
- Talking to ISO host asynchronously.

## [0.1.4] - 2020-04-27
### Added
- Deserializing http requests and sending them **synchronously** (i.e. blocking) to the hardcoded ISO host.

## [0.1.3] - 2020-04-26
### Added
- Using tokio and hyper frameworks to handle http requests.

## [0.1.2] - 2020-04-16
### Added
- ``iso_fields`` object as a part of ``struct Request``.

## [0.1.1] - 2020-04-16
### Added
- Dummy serialization from json to xml

## [0.1.0] - 2020-04-15
### Added
- Initial commit


src/dhi.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use serde_xml_rs::to_string;

use crate::errors::AppError;
use std::collections::BTreeMap;

mod util;

#[derive(Serialize, Debug)]
pub struct Header {
    #[serde(rename(serialize = "MessageID"))]
    message_id: i64,
    #[serde(rename(serialize = "SystemID"))]
    system_id: String,
}

#[derive(Serialize, Debug)]
#[serde(rename(serialize = "RequestInput"))]
pub struct DHIRequest {
    #[serde(rename(serialize = "ISO8583-87"))]
    iso_fields: Value,
}

impl DHIRequest {
    pub fn new(iso_obj: Value) -> Self {
        let mut req = DHIRequest {
            iso_fields: iso_obj,
        };
        if req.iso_fields.get("i007") == None {
            req.iso_fields["i007"] = serde_json::value::Value::String(util::get_mmddhhmmss());
        }
        if req.iso_fields.get("i011") == None {
            req.iso_fields["i011"] = serde_json::value::Value::String(util::gen_stan());
        }
        if req.iso_fields.get("i012") == None {
            req.iso_fields["i012"] = serde_json::value::Value::String(util::get_hhmmss());
        }
        if req.iso_fields.get("i013") == None {
            req.iso_fields["i013"] = serde_json::value::Value::String(util::get_mmdd());
        }
        if req.iso_fields.get("i037") == None {
            req.iso_fields["i037"] = serde_json::value::Value::String(util::gen_rrn());
        }
        req
    }

    pub fn serialize(&self) -> Result<String, AppError> {
        let serialized = to_string(&self)?;
        let serialized = format!("{:05}{}", serialized.len(), serialized);
        Ok(serialized)
    }
}

#[derive(Deserialize, Debug)]
#[serde(rename(deserialize = "Result"))]
#[serde(rename_all = "PascalCase")]
pub struct DHIResult {
    code: i32,
    description: String,
}

#[derive(Deserialize, Debug)]
#[serde(rename(deserialize = "RequestResponse"))]
pub struct DHIResponse {
    #[serde(rename(deserialize = "Result"))]
    res: DHIResult,

    #[serde(rename(deserialize = "ISO8583-87"))]
    iso_fields: BTreeMap<String, String>,
}

impl DHIResponse {
    pub fn serialize(&self) -> Result<String, AppError> {
        let serialized = serde_json::to_string(&self.iso_fields)?;
        Ok(serialized)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_xml_rs::from_reader;

    #[test]
    fn dhi_request_new_existing_fields() {
        let iso_data = r#"{
            "i000": "0100",
            "i002": "521324******0895",
            "i007": "Transmission date & time ",
            "i011": "STAN",
            "i012": "hhmmss",
            "i013": "MMDD",
            "i037": "Retrieval Reference Number"
        }"#;

        let r: DHIRequest = DHIRequest::new(serde_json::from_str(&iso_data).unwrap());
        assert_eq!(
            r.iso_fields["i007"], "Transmission date & time ",
            "i007 original value should be left untouched"
        );
        assert_eq!(
            r.iso_fields["i011"], "STAN",
            "i011 original value should be left untouched"
        );
        assert_eq!(
            r.iso_fields["i012"], "hhmmss",
            "i012 original value should be left untouched"
        );
        assert_eq!(
            r.iso_fields["i013"], "MMDD",
            "i013 original value should be left untouched"
        );
        assert_eq!(
            r.iso_fields["i037"], "Retrieval Reference Number",
            "Provided RRN should be left untouched"
        );
    }

    #[test]
    fn dhi_response_deserialization() {
        let s = r##"
        <?xml version="1.0"?>
        <RequestResponse>
            <Header/>
            <Result><Code>0</Code><Description>OK</Description></Result>
            <ISO8583-87><i000>0110</i000><i002>553691******0961</i002><i003>300000</i003><i004>000000000000</i004><i007>2804114717</i007><i043><![CDATA[IDDQD AM. 341215574     341215574 MSKRU]]></i043><i120>UD038IR0044444CR009ES0048100IA0103510198686</i120></ISO8583-87>"
        </RequestResponse>
        "##;

        let resp: DHIResponse = from_reader(s.as_bytes()).unwrap();

        assert_eq!(resp.res.code, 0);
        assert_eq!(resp.res.description, "OK");

        assert_eq!(resp.iso_fields["i000"], "0110");
        assert_eq!(resp.iso_fields["i002"], "553691******0961");
        assert_eq!(resp.iso_fields["i003"], "300000");
        assert_eq!(resp.iso_fields["i004"], "000000000000");
        assert_eq!(resp.iso_fields["i007"], "2804114717");
        assert_eq!(
            resp.iso_fields["i043"],
            "IDDQD AM. 341215574     341215574 MSKRU"
        );
        assert_eq!(
            resp.iso_fields["i120"],
            "UD038IR0044444CR009ES0048100IA0103510198686"
        );

        assert_eq!(resp.serialize().unwrap(), "{\"i000\":\"0110\",\"i002\":\"553691******0961\",\"i003\":\"300000\",\"i004\":\"000000000000\",\"i007\":\"2804114717\",\"i043\":\"IDDQD AM. 341215574     341215574 MSKRU\",\"i120\":\"UD038IR0044444CR009ES0048100IA0103510198686\"}");
    }
}
